packages/block-library/src/heading/edit.js

```diff
@@ -13,48 +13,30 @@ import HeadingToolbar from './heading-toolbar';
  */
 import { __ } from '@wordpress/i18n';
 import { PanelBody } from '@wordpress/components';
-import { compose } from '@wordpress/compose';
 import { createBlock } from '@wordpress/blocks';
 import {
 	AlignmentToolbar,
 	BlockControls,
 	InspectorControls,
 	RichText,
-	withColors,
-	PanelColorSettings,
+	__experimentalUseColors,
 } from '@wordpress/block-editor';
-import { memo } from '@wordpress/element';
-
-const HeadingColorUI = memo(
-	function( {
-		textColorValue,
-		setTextColor,
-	} ) {
-		return (
-			<PanelColorSettings
-				title={ __( 'Color Settings' ) }
-				initialOpen={ false }
-				colorSettings={ [
-					{
-						value: textColorValue,
-						onChange: setTextColor,
-						label: __( 'Text Color' ),
-					},
-				] }
-			/>
-		);
-	}
-);
 
 function HeadingEdit( {
 	attributes,
 	setAttributes,
 	mergeBlocks,
 	onReplace,
 	className,
-	textColor,
-	setTextColor,
 } ) {
+	const { TextColor, InspectorControlsColorPanel } = __experimentalUseColors(
+		[ { name: 'textColor', attribute: 'color' } ],
+		{
+			contrastCheckerProps: { backgroundColor: 'white', isLargeText: true },
```

Comment:
How do you decide which color from the first argument to use in the color contrast checks?

Reply:
It checks all of them.

Reply:
I think we should not check the contrast of all the colors. Imagining a scenario where we use the hook with background color and text color, the contrast checked should not check the background color.
Another point is that I think we should not hardcode a color for contrast checking (currently we assume the background is always white). Themes may set the heading background to block, and in that case, white text color is perfectly fine.
I think users should be able to configure the color pairs used in contrast checking.
We currently also have some mechanisms to retrieve colors using browser computed styles functions when they are not set. This mechanism will need to be improved, e.g., it was created before nesting existed and didn't handle it. As an example of the imporvement needed for nesting, If we have a heading with black color inside a group block with block background color if the theme does not explicitly set a heading background color we should show a contrast warning.
Having all contrast functionality in this hook makes it very huge. I think the contrast checking should be outside of this hook, going in the direction of what @mcsf referred (multiple hooks) the developers use useColor for the color logic useContrast for the contrast logic.
The contrast hook can contain the logic of the fallback styles we currently have, and receive some color values retrieved from useColors and simply returns a warning message or null.

Reply:
>I think we should not check the contrast of all the colors. Imagining a scenario where we use the hook with background color and text color, the contrast checked should not check the background color.

They can separate it into two hook calls or provide their own custom `ConstrastChecker` child.

>Another point is that I think we should not hardcode a color for contrast checking (currently we assume the background is always white). Themes may set the heading background to block, and in that case, white text color is perfectly fine.

Nothing stops that `backgroundColor` property from being a variable.

>This mechanism will need to be improved, e.g., it was created before nesting existed and didn't handle it. As an example of the imporvement needed for nesting, If we have a heading with black color inside a group block with block background color if the theme does not explicitly set a heading background color we should show a contrast warning.

I think that sort of contrast checking behavior belongs in another hook in another PR.